PY-20075/PY-28393/PY-28332: Correctly resolve PyModuleType

fitermay opened the pull request.

fitermay commented on python/src/com/jetbrains/python/psi/impl/PyFileImpl.java:
```diff
@@ -409,14 +409,16 @@ else if (element != null && element.getContainingFile() == this) {
         .resolveModuleAt(QualifiedName.fromComponents(name), getContainingDirectory(), PyResolveImportUtil.fromFoothold(this))
         .forEach(module -> allFallbackResults.poke(module, RatedResolveResult.RATE_LOW));
 
+      final PsiElement allElement = findExportedName(PyNames.ALL);
```
@vlasovskikh This change was added to avoid breaking an existing test.

vlasovskikh replied:
@fitermay Found it, thanks.

fitermay commented on python/src/com/jetbrains/python/psi/resolve/ResolveImportUtil.java:
```diff
@@ -275,7 +274,7 @@ else if (parent instanceof PsiDirectory) {
       if (!fileOnly || PsiTreeUtil.instanceOf(moduleMember, PsiFile.class, PsiDirectory.class)) {
         results.add(member);
-        if (moduleMember != null && !preferResolveInDirectoryOverModule(moduleMember)) {
+        if (moduleMember != null && (!preferResolveInDirectoryOverModule(moduleMember) || !PyUtil.isPackage(parent))) {
```
@vlasovskikh This change was added to avoid breaking an existing test.

vlasovskikh replied:
@fitermay I've commented it out and no test has failed. Are you sure about this one?

fitermay replied:
@vlasovskikh You're right.  Another change down the line must have covered it.

fitermay commented on python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java:
```diff
@@ -40,23 +34,39 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
+        continue;
+      }
+      //if it's a non-package module or a synthetic import element
+      if (!PyUtil.isPackage(resolvedModuleOrPackage, location) || myImportedModule.getImportElement() == null) {
```
@vlasovskikh This change was added to avoid breaking an existing tests. This is to resolve synthetic imports like os.path

vlasovskikh replied:
@fitermay I've commented out `!PyUtil.isPackage(resolvedModuleOrPackage, location)` and no test has failed. I'm fine with `myImportedModule.getImportElement() == null` which is necessary indeed.

fitermay replied:
You're right. In cases where we need to resolve inside 'fake' packages like `os` myImportedModule.getImportElement() is always null so it covers those cases.

Could you please elaborate on the comment : 'Non-package modules cannot have file system based children'?    Just to improve my own understanding...

vlasovskikh replied:
@fitermay For `ResolveImportUtil.resolveChildren()` which you call here a package is either a directory with `__init__.py` or a 
Python 3.3+ namespace directory. But both of these things are packages according to `PyUtil.isPackage()`, so this code won't be executed under anyway. Regular Python files are not packages, so it doesn't make sense to resolve their children by searching them as files on a file system.

There is also the third case in `ResolveImportUtil.resolveChildren()`. When `parent` is neither of those things, then this method asks reference providers for a type of a resolved reference targets and then delegates resolving the name to it. But I suspect that this code is actually no longer used, so it makes no sense to call `resolveChildern()` just to invoke these reference providers.

fitermay commented on python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java:
```diff
@@ -40,23 +34,39 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
-          final Set<PsiElement> imported = Sets.newHashSet(importedSubmodules);
-          elements = ContainerUtil.filter(elements, element -> imported.contains(element));
+
+    List<PsiElement> importedModuleCandidates = myImportedModule.multiResolve();
```
Sure, I will take a look at it

fitermay commented on python/src/com/jetbrains/python/psi/impl/PyImportedModule.java:
```diff
@@ -123,4 +143,20 @@ private static PsiElement resolve(PyImportElement importElement, @NotNull final
   }
+
+  @Override
+  public boolean equals(Object o) {
```
That is one reason. More generally not having hashCode and equals in a LightPsiElement breaks the assumption most of the code makes about PsiElements not being duplicated when added to a set.  For example, without it TypeEvalContext will not cache the resolved value correctly.

fitermay replied:
I will try formulating a test for this.

fitermay commented on python/src/com/jetbrains/python/psi/impl/PyImportedModule.java:
```diff
@@ -111,6 +116,21 @@ public PsiElement resolve() {
   }
 
+  @NotNull
+  public List<PsiElement> multiResolve() {
```
Will do

sproshev replied:
I suggest not to drop rate and return `List<RatedResolveResult>` as in other places

fitermay commented on python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java:
```diff
@@ -40,23 +34,39 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
+      }
+      //if it's a non-package module or a synthetic import element
+      if (!PyUtil.isPackage(resolvedModuleOrPackage, location) || myImportedModule.getImportElement() == null) {
+        PsiFile resolvingFromFile = Optional.ofNullable(location).map(l -> l.getContainingFile()).orElse(null);
```
Yes, there is too much ceremony here  with Optional.

sproshev commented on python/src/com/jetbrains/python/psi/PyUtil.java:
```diff
@@ -549,7 +549,10 @@ public static PsiElement resolveToTheTop(@NotNull final PsiElement elementToReso
       }
       final PsiElement resolve = reference.resolve();
-      if ((resolve == null) ||  checkedElements.contains(resolve) || resolve.equals(currentElement) || !inSameFile(resolve, currentElement)) {
+      if ((resolve == null) ||
```
I'd remove unnecessary parentheses here so the condition will be left as one line

sproshev commented on python/src/com/jetbrains/python/psi/impl/PyImportedModule.java:
```diff
@@ -111,6 +116,21 @@ public PsiElement resolve() {
+    }
+
+    return results.stream().map(res -> res.getElement())
+                  .map((PsiElement el) -> el instanceof PsiDirectory ? PyUtil.getPackageElement((PsiDirectory)el, this) : el)
```
No need to specify type

sproshev commented on python/src/com/jetbrains/python/psi/impl/PyImportedModule.java:
```diff
@@ -111,6 +116,21 @@ public PsiElement resolve() {
+
+    return results.stream().map(res -> res.getElement())
+                  .map((PsiElement el) -> el instanceof PsiDirectory ? PyUtil.getPackageElement((PsiDirectory)el, this) : el)
+                  .collect(Collectors.toList());
```
Consider using `StreamEx` library, it has useful methods for streams.
E.g. `collect(Collectors.toList())` could be rewritten as `toList()`.

sproshev commented on python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java:
```diff
@@ -40,23 +34,39 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
+          fallbackResults.stream()
+                         .map(res -> new ImportedResolveResult(res.getElement(), res.getRate(), myImportedModule.getImportElement()))
+                         .forEach(ret::add);
+          return ret;
```
Consider using our `ContainerUtil` classes
E.g. this code could be rewritten as `return ContainerUtil.map(fallbackResults, res -> new ImportedResolveResult(res.getElement(), res.getRate(), myImportedModule.getImportElement()));`

fitermay commented:
Thanks for the comments. I will make the requested changes.

fitermay commented:
@vlasovskikh @sproshev 
I've implemented most of the changes and fixed a couple more issues. What's outstanding is:

1.  Figuring out how to create a high level test for the equals and hashCode change. I've made that change not to fix a specific behavior but out of general principle

2. Removing usages of deprecated methods. It is not always obvious what to do in every case. So this may take a while

3. `collectImportedSubmodules` and `resolveImplicitPackageMember` are now very similar. So that may be worth refactoring...

vlasovskikh commented on python/src/com/jetbrains/python/psi/resolve/ResolveImportUtil.java:
```diff
@@ -191,14 +192,27 @@ public static PsiElement resolveImportElement(PyImportElement importElement, @No
+  @NotNull
+  public static List<PsiElement> multiResolveModuleInRoots(@NotNull QualifiedName moduleQualifiedName, @Nullable PsiElement foothold) {
+    if (foothold == null) return Collections.emptyList();
+    return PyResolveImportUtil.resolveQualifiedName(moduleQualifiedName,
```
This call has weird indentation levels.

vlasovskikh commented on python/src/com/jetbrains/python/psi/resolve/ResolveImportUtil.java:
```diff
@@ -191,14 +192,27 @@ public static PsiElement resolveImportElement(PyImportElement importElement, @No
-                                                                              PyResolveImportUtil.fromFoothold(foothold));
-    return !results.isEmpty() ? results.get(0) : null;
+    final List<PsiElement> results = multiResolveModuleInRoots(moduleQualifiedName, foothold);
+    if (results.isEmpty()) {
```
A shorter form: `return ContainerUtil.getFirstItem(results);`

vlasovskikh commented on python/src/com/jetbrains/python/psi/types/PyModuleType.java:
```diff
@@ -137,20 +161,23 @@ private static PsiElement resolveByOverridingMembersProviders(@NotNull PyFile mo
-  private List<? extends RatedResolveResult> resolveImplicitPackageMember(@NotNull String name,
-                                                                          @NotNull List<PyImportElement> importElements) {
-    final VirtualFile moduleFile = myModule.getVirtualFile();
+  private static List<? extends RatedResolveResult> resolveImplicitPackageMember(@NotNull PsiFileSystemItem moduleOrPackage,
```
I agree, `resolveImplicitPackageMember` and `collectImportedSubmodules` do look similar. Not sure how to refactor them to remove this duplication though. Feel free to skip it if you're not sure too (a TODO item might be OK for now).

vlasovskikh commented:
I've come up with a test for equality:

```
root1/
  nspkg1/
    m1.py
root2/
  nspkg1/
    m1.py
a.py
```

a.py:

```python
from nspkg1 import not_found  # Error: Cannot find reference 'not_found' in 'imported module nspkg1 | imported module nspkg1'
```

We have a union in the text message even though it's the same imported module. To fix this you'll have to introduce `equals` and `hashCode` for `PyImportedModuleType` as well. See `PyUnionType.myMembers`, `com.jetbrains.python.inspections.Py3UnresolvedReferencesInspectionTest#doMultiFileTest(java.lang.String, java.util.List<java.lang.String>)`.

vlasovskikh commented:
@fitermay Could you please fix the issue with `equals` and `hashCode` and optionally address some code style comments? I'll be happy to accept this PR after the fix.

sproshev commented on python/src/com/jetbrains/python/psi/impl/PyImportedModule.java:
```diff
@@ -96,19 +103,33 @@ public PyImportElement getImportElement() {
-    }
-    return element;
+    return StreamEx.of(results).map(this::tryReplaceDirWithPackage)
+                   .toList();
```
`com.intellij.util.containers.ContainerUtil#map`

sproshev commented on python/src/com/jetbrains/python/psi/impl/ResolveResultList.java:
```diff
@@ -37,6 +43,19 @@
+    if (from == null || from.isEmpty()) {
+      return Collections.emptyList();
+    }
+    return StreamEx.of(from).map(res -> new ImportedResolveResult(res.getElement(), res.getRate(), nameDefiner)).toList();
```
`com.intellij.util.containers.ContainerUtil#map`

fitermay commented:
@vlasovskikh @sproshev Thanks for the review. I've applied the changes

vlasovskikh commented:
@fitermay I'm afraid one of the tests has become red after the recent changes in code completion: `com.jetbrains.python.PythonCompletionTest#testSubmoduleOfIndirectlyImportedPackage`. Could you please take a look?

fitermay commented:
Could you please rerun the test? It seems to fail sporadically.

On Fri, Feb 9, 2018 at 10:56 AM Andrey Vlasovskikh <notifications@github.com>
wrote:

> @fitermay <https://github.com/fitermay> I'm afraid one of the tests has
> become red after the recent changes in code completion:
> com.jetbrains.python.PythonCompletionTest#testSubmoduleOfIndirectlyImportedPackage.
> Could you please take a look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JetBrains/intellij-community/pull/728#issuecomment-364473846>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJZNNYXu_TsvsJucMm_XZx6DdZ9a9v5Gks5tTGqWgaJpZM4R968s>
> .
>

fitermay commented:
I will take a look later tonight though. The non deterministic behavior
itself is interesting:)

On Fri, Feb 9, 2018 at 11:16 AM Yuli Fiterman <fiterman@gmail.com> wrote:

> Could you please rerun the test? It seems to fail sporadically.
>
> On Fri, Feb 9, 2018 at 10:56 AM Andrey Vlasovskikh <
> notifications@github.com> wrote:
>
>> @fitermay <https://github.com/fitermay> I'm afraid one of the tests has
>> become red after the recent changes in code completion:
>> com.jetbrains.python.PythonCompletionTest#testSubmoduleOfIndirectlyImportedPackage.
>> Could you please take a look?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JetBrains/intellij-community/pull/728#issuecomment-364473846>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AJZNNYXu_TsvsJucMm_XZx6DdZ9a9v5Gks5tTGqWgaJpZM4R968s>
>> .
>>
>

vlasovskikh commented:
@fitermay It failed for me several times in a row. I have to tell that I was running it with other tests when it failed.

fitermay commented:
@vlasovskikh 

You can comment out tests `testExcludedSubPackage()` and `testExcludedSubPackage` and the test will pass.  Apparently `PsiTestUtil.addExcludedRoot(...)` stays even after the tests are finished and that adds pkg1 to excluded roots!

fitermay commented:
@vlasovskikh I've fixed the tear down issue in `PythonCompletionTest` as well as make a few more minor changes

vlasovskikh commented:
@fitermay Thanks for your fixes! I've rebased your PR onto master.
